pkg/flow/internal/importsource/import_file.go

```diff
@@ -3,96 +3,236 @@ package importsource
+		}
+	}
+
+	var err error
+	switch im.args.Type {
+	case filedetector.DetectorPoll:
+		im.detector = filedetector.NewPoller(filedetector.PollerOptions{
+			Filename:      im.args.Filename,
+			ReloadFile:    reloadFile,
+			PollFrequency: im.args.PollFrequency,
+		})
+	case filedetector.DetectorFSNotify:
+		im.detector, err = filedetector.NewFSNotify(filedetector.FsNotifyOptions{
+			Logger:        im.managedOpts.Logger,
+			Filename:      im.args.Filename,
+			ReloadFile:    reloadFile,
+			PollFrequency: im.args.PollFrequency,
+		})
+	}
+
+	return err
 }
 
 func (im *ImportFile) Run(ctx context.Context) error {
-	return im.fileComponent.Run(ctx)
+	defer func() {
+		if err := im.detector.Close(); err != nil {
+			level.Error(im.managedOpts.Logger).Log("msg", "failed to shut down detector", "err", err)
+		}
+		im.detector = nil
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-im.reloadCh:
+			time.Sleep(waitReadPeriod)
+
+			// We ignore the error here from readFile since readFile will log errors
+			// and also report the error and update the health of the source.
+			_ = im.readFile()
+		}
+	}
+}
+
+func (im *ImportFile) readFile() error {
+	files, err := im.collectFiles()
+	if err != nil {
+		im.setHealth(component.Health{
+			Health:     component.HealthTypeUnhealthy,
+			Message:    fmt.Sprintf("failed to collect files: %s", err),
+			UpdateTime: time.Now(),
+		})
+		level.Error(im.managedOpts.Logger).Log("msg", "failed to collect files", "err", err)
+		return err
+	}
+	fileContents := make(map[string]string)
+	for _, f := range files {
+		fpath := f
+		if im.importDir {
+			fpath = filepath.Join(im.args.Filename, fpath)
+		}
+		bb, err := os.ReadFile(fpath)
+		if err != nil {
+			im.setHealth(component.Health{
+				Health:     component.HealthTypeUnhealthy,
+				Message:    fmt.Sprintf("failed to read file: %s", err),
+				UpdateTime: time.Now(),
+			})
+			level.Error(im.managedOpts.Logger).Log("msg", "failed to read file", "file", fpath, "err", err)
+			return err
+		}
+		fileContents[f] = string(bb)
+	}
+
+	im.setHealth(component.Health{
+		Health:     component.HealthTypeHealthy,
+		Message:    "read file",
+		UpdateTime: time.Now(),
+	})
+	im.onContentChange(fileContents)
+	return nil
 }
 
-// CurrentHealth returns the health of the file component.
 func (im *ImportFile) CurrentHealth() component.Health {
-	return im.fileComponent.CurrentHealth()
+	im.healthMut.RLock()
+	defer im.healthMut.RUnlock()
+	return im.health
+}
+
+func (im *ImportFile) setHealth(h component.Health) {
+	im.healthMut.Lock()
+	defer im.healthMut.Unlock()
+	im.health = h
+}
+
+func (im *ImportFile) collectFiles() ([]string, error) {
+	fpath := im.args.Filename
+	fi, err := os.Stat(fpath)
+	if err != nil {
+		return nil, err
+	}
+
+	files := make([]string, 0)
+	im.importDir = fi.IsDir()
+	if im.importDir {
+		files, err = collectFilesFromDir(fpath)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		files = append(files, fpath)
+	}
```

Comment:
nit: it doesn't seem necessary to store `importDir` on the struct, since it's only ever written to and read on the same code path (`readFile`). 

You could change `collectFiles` to return whether it was a directory instead:

```go
func (im *ImportFile) collectFiles() (content []string, dir bool, err error)
```